ext/mbstring/mbstring.c

```diff
@@ -3183,77 +3183,368 @@ static int *make_conversion_map(HashTable *target_hash, int *convmap_size)
+		uint32_t offset  = mapelm[2];
+		uint32_t codepoint = number - offset;
+		if (codepoint >= lo_code && codepoint <= hi_code) {
+			*retval = codepoint;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static zend_string* html_numeric_entity_decode(zend_string *input, const mbfl_encoding *encoding, uint32_t *convmap, int mapsize)
+{
+	uint32_t wchar_buf[128], converted_buf[128];
+
+	unsigned int state = 0;
+	unsigned char *in = (unsigned char*)ZSTR_VAL(input);
+	size_t in_len = ZSTR_LEN(input);
+
+	mb_convert_buf buf;
+	mb_convert_buf_init(&buf, in_len, MBSTRG(current_filter_illegal_substchar), MBSTRG(current_filter_illegal_mode));
+
+	/* Decode input string from bytes to wchars one 128-wchar buffer at a time, then deconvert HTML entities,
+	 * copying the deconverted wchars to a second buffer, then convert back to original encoding from the
+	 * 2nd 'converted' buffer.
+	 *
+	 * Tricky part: an HTML entity might be truncated at the end of the wchar buffer; the remaining
+	 * part could come in the next buffer of wchars. To deal with this problem, when we find what looks
+	 * like an HTML entity, we scan to see if it terminates before the end of the wchar buffer or not.
+	 * If not, we copy it to the beginning of the wchar buffer, and tell the input conversion routine
+	 * to store the next batch of wchars after it.
+	 *
+	 * Optimization: Scan for &, and if we don't find it anywhere, don't even bother copying the
+	 * wchars from the 1st buffer to the 2nd one.
+	 *
+	 * 'converted_buf' is big enough that the deconverted wchars will *always* fit in it, so we don't
+	 * have to do bounds checks when writing wchars into it.
+	 */
+
+	unsigned int wchar_buf_offset = 0;
+
+	while (in_len) {
+		/* Leave space for sentinel at the end of the buffer */
+		size_t out_len = encoding->to_wchar(&in, &in_len, wchar_buf + wchar_buf_offset, 127 - wchar_buf_offset, &state);
+		out_len += wchar_buf_offset;
+		ZEND_ASSERT(out_len <= 127);
+		wchar_buf[out_len] = '&'; /* Sentinel */
+
+		uint32_t *p, *converted;
+
+		/* Scan for & first
+		 * If wchar_buf_offset > 0, then definitely & will be there, so don't bother in that case */
+		if (wchar_buf_offset == 0) {
+			p = wchar_buf;
+			while (*p != '&')
+				p++;
+			if (p == wchar_buf + out_len) {
+				/* No HTML entities in this buffer */
+				encoding->from_wchar(wchar_buf, out_len, &buf, !in_len);
+				continue;
+			} else {
+				/* Copy over the prefix with no & which we already scanned */
+				memcpy(converted_buf, wchar_buf, (p - wchar_buf) * 4);
+				converted = converted_buf + (p - wchar_buf);
+			}
+		} else {
+			p = wchar_buf;
+			converted = converted_buf;
+		}
+
+found_ampersand:
+		ZEND_ASSERT(*p == '&');
+		uint32_t *p2 = p;
+
+		if (*++p2 == '#') {
+			if (*++p2 == 'x') {
+				/* Possible hex entity */
+				uint32_t w = *++p2;
+				while ((w >= '0' && w <= '9') || (w >= 'A' && w <= 'F') || (w >= 'a' && w <= 'f'))
+					w = *++p2;
+				if (p2 == wchar_buf + out_len) {
```

Comment:
```
uint32_t w = *++p2;
while ((w >= '0' && w <= '9') || (w >= 'A' && w <= 'F') || (w >= 'a' && w <= 'f'))
	w = *++p2;
if (p2 == wchar_buf + out_len) {
	/* We hit the end of the buffer while still reading digits */
	if ((p2 - p) <= 11 && in_len) {
		/* The number of digits was legal (1-8 hex digits)
		 * We need to process this identity again on the next iteration of the main loop */
		memmove(wchar_buf, p, (p2 - p) * 4);
		wchar_buf_offset = p2 - p;
		goto process_converted_wchars;
	} else {
		/* Invalid entity */
		memcpy(converted, p, (p2 - p) * 4);
		converted += p2 - p;
		goto process_converted_wchars_no_offset;
	}

```

If the `while` loop consumes hex digits until it hits the sentinel at the end of the wchar buffer, then `if (p2 == wchar_buf + out_len)` will be true. If it has read no more than 8 hex digits, meaning the entity *could* be legal (perhaps the first wchar in the next buffer will be `;`), it will copy `&#x<DIGITS>` to the beginning of the wchar buffer and refill the buffer, making sure not to overwrite them.

But if it's at the very end of the input string, it can't refill the wchar buffer, so in that case it will just copy `&#x` to the output and terminate. Same if the number of digits read was too high for the entity to be valid.

Reply:
Maybe I should also say... `p` will point to `&`, `p2` will point to the next character *after* the last digit. But since we copy only `p2 - p` characters through to the output if we discover the entity is invalid, the one which `p2` points to will **not** be copied through.